pkg/gatewayserver/io/ws/sessions.go

```diff
@@ -0,0 +1,43 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ws
+
+import (
+	"context"
+	"sync/atomic"
+)
+
+type sessionKeyType struct{}
+
+var sessionKey sessionKeyType
+
+// Session contains the session state for a single gateway.
+type Session struct {
+	State atomic.Value
+}
```

Comment:
Either don't wrap `atomic.Value` in a struct, or declare a `Data interface{}` and let the consumer figure out what to put in there. I prefer the latter

Reply:
Ok but this does something similar right? If I use `Data interface{}` then I'll need to add a mutex as well to `Session` since UL/DL paths can write to this concurrently.

Reply:
I meant that the caller then assumes what type `Data interface{}` is, if the caller uses it from multiple goroutines and needs protection